### ISSUE TITLE
Gitlab install fails under [WSL docker-interop-client] windows's docker with host volumes

### ISSUE THREAD

**Issue by n3rd4i** [microsoft/WSL#3190]
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)
`Microsoft Windows [Version 10.0.17134.48]`

* Docker and WSL distro versions:
`WSL Description:	Ubuntu 18.04 LTS`
`Docker version 18.05.0-ce-rc1, build 33f00ce` -- windows docker version
`Docker version 18.04.0-ce, build 3d479c0` -- wsl [inter-op docker.io](https://blogs.msdn.microsoft.com/commandline/2017/12/08/cross-post-wsl-interoperability-with-docker/) version
* **wsl.conf**
```
[automount]
enabled = true
options = "metadata"
```
* `mount -l`
```
rootfs on / type lxfs (rw,noatime)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,noatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,noatime)
none on /dev type tmpfs (rw,noatime,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,noatime,gid=5,mode=620)
none on /run type tmpfs (rw,nosuid,noexec,noatime,mode=755)
none on /run/lock type tmpfs (rw,nosuid,nodev,noexec,noatime)
none on /run/shm type tmpfs (rw,nosuid,nodev,noatime)
none on /run/user type tmpfs (rw,nosuid,nodev,noexec,noatime,mode=755)
binfmt_misc on /proc/sys/fs/binfmt_misc type binfmt_misc (rw,noatime)
C: on /mnt/c type drvfs (rw,noatime,uid=1000,gid=1000,metadata)

```

* What you're doing and what's happening:
`docker pull --platform=linux gitlab/gitlab-ce`
`mkdir -p config;mkdir -p logs;mkdir -p data;`
`strace -f -o docker_gitlab_wsl.strace docker run --platform=linux --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v c:/gitlab-ce/config:/etc/gitlab -v c:/gitlab-ce/logs:/var/log/gitlab -v c:/gitlab-ce/data:/var/opt/gitlab --rm gitlab/gitlab-ce >docker.log 2>&1`

* What's wrong / what should be happening instead:  (Expected command should be successful)

```
[...]
Running handlers:
There was an error running gitlab-ctl reconfigure:

storage_directory[/var/opt/gitlab/.ssh] (gitlab::gitlab-shell line 38) had an error: Mixlib::ShellOut::ShellCommandFailed: ruby_block[directory resource: /var/opt/gitlab/.ssh] (/opt/gitlab/embedded/cookbooks/cache/cookbooks/package/resources/storage_directory.rb line 33) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of chmod 00700 /var/opt/gitlab/.ssh ----
STDOUT: 
STDERR: chmod: changing permissions of '/var/opt/gitlab/.ssh': Operation not permitted
---- End output of chmod 00700 /var/opt/gitlab/.ssh ----
Ran chmod 00700 /var/opt/gitlab/.ssh returned 1

Running handlers complete
Chef Client failed. 
[...]
```

* Strace of the failing command, in a [gist](https://gist.github.com/n3rd4i/5493df264ce90ef1889026ea6f5cdd49)

* docker output full log [gist](https://gist.github.com/n3rd4i/29ba9e14b44e0fc00e0092cd0c8ee39d)

See [our contributing instructions](https://github.com/Microsoft/BashOnWindows/blob/master/CONTRIBUTING.md) for assistance.

**Comment by therealkenc** [comment]
Strange thing is I can't match that descriptor 3 being `read()` below. Last time (unrelated) fd 3 was heard of it was `close()d` [here](https://gist.github.com/n3rd4i/5493df264ce90ef1889026ea6f5cdd49#file-docker_gitlab_wsl-strace-L433). Do the `strace` again with `-f -o strace -o docker_gitlab_wsl.strace docker`. At minimum there are `clone()s` [in there](https://gist.github.com/n3rd4i/5493df264ce90ef1889026ea6f5cdd49#file-docker_gitlab_wsl-strace-L213) but no PIDs in your `strace` output. I don't think that will explain fd 3 but at least it will be complete.

As a general observation, `pselect(...) = TIMEOUT` seems to be a common final nail in the coffin lately, even if that syscall itself isn't necessary the root cause.

```
read(3, "HTTP/1.1 201 Created\r\nApi-Versio"..., 4096) = 305
futex(0xc42048e538, FUTEX_WAKE, 1)      = 1
clock_gettime(CLOCK_REALTIME, {tv_sec=1526235095, tv_nsec=215657400}) = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=144, tv_nsec=886688000}) = 0
futex(0xc4204b8138, FUTEX_WAKE, 1)      = 1
read(3, 0xc4202ae000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
read(3, 0xc4202ae000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
futex(0x256da18, FUTEX_WAIT, 0, NULL)   = 0
futex(0xc4204b8138, FUTEX_WAKE, 1)      = 1
futex(0x256da18, FUTEX_WAIT, 0, NULL)   = 0
futex(0x256da18, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
pselect6(0, NULL, NULL, NULL, {tv_sec=0, tv_nsec=3000}, NULL) = 0 (Timeout)
futex(0x256da18, FUTEX_WAIT, 0, NULL)   = -1 EAGAIN (Resource temporarily unavailable)
pselect6(0, NULL, NULL, NULL, {tv_sec=0, tv_nsec=3000}, NULL) = 0 (Timeout)
```

**Comment by n3rd4i** [comment]
I've updated the docker_gitlab_wsl.strace on gist to include trace for child processes.

**Comment by n3rd4i** [comment]
For comparison, after more investigation, I noticed that there are 4 scenarios in total:
1. Docker windows containers mode (within WSL) with host volumes (**fails** with above description)
`docker run --platform=linux --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v c:/gitlab-ce/config:/etc/gitlab -v c:/gitlab-ce/logs:/var/log/gitlab -v c:/gitlab-ce/data:/var/opt/gitlab --rm gitlab/gitlab-ce`
2. Docker windows containers mode (within WSL) with container volumes (**fails** with above description)
`docker run --platform=linux --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v gitlab_config:/etc/gitlab -v gitlab_logs:/var/log/gitlab -v gitlab_data:/var/opt/gitlab --rm gitlab/gitlab-ce`
3. Docker linux containers mode (within Hyper-V) with host volumes (**fails** with above description)
`docker run --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v c:/gitlab-ce/config:/etc/gitlab -v c:/gitlab-ce/logs:/var/log/gitlab -v c:/gitlab-ce/data:/var/opt/gitlab --rm gitlab/gitlab-ce`
4. Docker linux containers mode (within Hyper-V) with container volumes (**passes**, the container is successful)
`docker run --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v gitlab_config:/etc/gitlab -v gitlab_logs:/var/log/gitlab -v gitlab_data:/var/opt/gitlab --rm gitlab/gitlab-ce`

The result files of failed cases are similar so I did not attach them.

**Comment by n3rd4i** [comment]
**Describe the bug**
Re-run the command #2 with more recent versions:

**To Reproduce**
Steps to reproduce the behavior:
0. Previous requirements: Docker installed, WSL `docker-relay` installed and working
1. Run `docker run --platform=linux --name gitlab -p 6080:80 -p 6022:22 -p 6443:443 -v gitlab_config:/etc/gitlab -v gitlab_logs:/var/log/gitlab -v gitlab_data:/var/opt/gitlab --rm gitlab/gitlab-ce`
4. See error log:
``` log
Thank you for using GitLab Docker Image!
Current version: gitlab-ce=11.6.1-ce.0

Configure GitLab for your system by editing /etc/gitlab/gitlab.rb file
And restart this container to reload settings.
To do it use docker exec:

  docker exec -it gitlab vim /etc/gitlab/gitlab.rb
  docker restart gitlab

For a comprehensive list of configuration options please see the Omnibus GitLab readme
https://gitlab.com/gitlab-org/omnibus-gitlab/blob/master/README.md

If this container fails to start due to permission problems try to fix it by executing:

  docker exec -it gitlab update-permissions
  docker restart gitlab

Preparing services...
Starting services...
Configuring GitLab...
/opt/gitlab/embedded/bin/runsvdir-start: line 24: ulimit: pending signals: cannot modify limit: Operation not permitted
/opt/gitlab/embedded/bin/runsvdir-start: line 29: ulimit: open files: cannot modify limit: Operation not permitted
/opt/gitlab/embedded/bin/runsvdir-start: line 34: ulimit: max user processes: cannot modify limit: Operation not permitted
/opt/gitlab/embedded/bin/runsvdir-start: line 37: /proc/sys/fs/file-max: Read-only file system
Starting Chef Client, version 13.6.4
resolving cookbooks for run list: ["gitlab"]
Synchronizing Cookbooks:
  - gitlab (0.0.1)
  - package (0.1.0)
  - postgresql (0.1.0)
  - redis (0.1.0)
  - registry (0.1.0)
  - mattermost (0.1.0)
  - consul (0.0.0)
  - gitaly (0.1.0)
  - letsencrypt (0.1.0)
  - nginx (0.1.0)
  - runit (0.14.2)
  - acme (3.1.0)
  - crond (0.1.0)
  - compat_resource (12.19.0)
Installing Cookbook Gems:
Compiling Cookbooks...
Recipe: gitlab::default
  * directory[/etc/gitlab] action create (up to date)
  Converging 486 resources
  * directory[/etc/gitlab] action create (up to date)
  * directory[Create /var/opt/gitlab] action create (up to date)
  * directory[/opt/gitlab/embedded/etc] action create
    - create new directory /opt/gitlab/embedded/etc
    - change mode from '' to '0755'
    - change owner from '' to 'root'
    - change group from '' to 'root'
  * template[/opt/gitlab/embedded/etc/gitconfig] action create
    - create new file /opt/gitlab/embedded/etc/gitconfig
    - update content in file /opt/gitlab/embedded/etc/gitconfig from none to 987af3
    --- /opt/gitlab/embedded/etc/gitconfig	2019-01-08 20:26:21.070000000 +0000
    +++ /opt/gitlab/embedded/etc/.chef-gitconfig20190108-30-13m9x7c	2019-01-08 20:26:21.070000000 +0000
    @@ -1 +1,11 @@
    +[pack]
    +  threads = 1
    +[receive]
    +  fsckObjects = true
    +advertisePushOptions = true
    +[repack]
    +  writeBitmaps = true
    +[transfer]
    +  hideRefs=^refs/tmp/
    +hideRefs=^refs/keep-around/
    - change mode from '' to '0755'
Recipe: gitlab::web-server
  * account[Webserver user and group] action create (up to date)
Recipe: gitlab::users
  * directory[/var/opt/gitlab] action create (up to date)
  * account[GitLab user and group] action create (up to date)
  * template[/var/opt/gitlab/.gitconfig] action create
    - update content in file /var/opt/gitlab/.gitconfig from a2fa1e to 3a6fef
    --- /var/opt/gitlab/.gitconfig	2019-01-08 20:24:21.563018900 +0000
    +++ /var/opt/gitlab/.chef-.gitconfig20190108-30-675lz5.gitconfig	2019-01-08 20:26:21.653416900 +0000
    @@ -4,7 +4,7 @@
     
     [user]
             name = GitLab
    -        email = gitlab@9187d868e738
    +        email = gitlab@67be2e823378
     [core]
             autocrlf = input
     [gc]
  * directory[/var/opt/gitlab/.bundle] action create (up to date)
Recipe: gitlab::gitlab-shell
  * storage_directory[/var/opt/gitlab/.ssh] action create
    * ruby_block[directory resource: /var/opt/gitlab/.ssh] action run (skipped due to not_if)
     (up to date)
  * directory[/var/log/gitlab/gitlab-shell/] action create (up to date)
  * directory[/var/opt/gitlab/gitlab-shell] action create (up to date)
  * templatesymlink[Create a config.yml and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-shell/config.yml] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-shell/config.yml to /var/opt/gitlab/gitlab-shell/config.yml] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-shell/config.yml to /var/opt/gitlab/gitlab-shell/config.yml
  
  * link[/opt/gitlab/embedded/service/gitlab-shell/.gitlab_shell_secret] action create
    - create symlink at /opt/gitlab/embedded/service/gitlab-shell/.gitlab_shell_secret to /opt/gitlab/embedded/service/gitlab-rails/.gitlab_shell_secret
  * execute[/opt/gitlab/embedded/service/gitlab-shell/bin/gitlab-keys check-permissions] action run
    - execute /opt/gitlab/embedded/service/gitlab-shell/bin/gitlab-keys check-permissions
  * bash[Set proper security context on ssh files for selinux] action run (skipped due to only_if)
Recipe: gitlab::gitlab-rails
  * storage_directory[/var/opt/gitlab/git-data] action create
    * ruby_block[directory resource: /var/opt/gitlab/git-data] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/git-data/repositories] action create
    * ruby_block[directory resource: /var/opt/gitlab/git-data/repositories] action run (skipped due to not_if)
     (up to date)
  * directory[/var/log/gitlab] action create (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/artifacts] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/artifacts] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/lfs-objects] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/lfs-objects] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/packages] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/packages] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/uploads] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/uploads] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-ci/builds] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-ci/builds] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/cache] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/cache] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/tmp] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/tmp] action run (skipped due to not_if)
     (up to date)
  * storage_directory[/var/opt/gitlab/gitlab-rails/shared/pages] action create
    * ruby_block[directory resource: /var/opt/gitlab/gitlab-rails/shared/pages] action run (skipped due to not_if)
     (up to date)
  * directory[create /var/opt/gitlab/gitlab-rails/etc] action create (up to date)
  * directory[create /opt/gitlab/etc/gitlab-rails] action create
    - create new directory /opt/gitlab/etc/gitlab-rails
    - change mode from '' to '0700'
    - change owner from '' to 'git'
  * directory[create /var/opt/gitlab/gitlab-rails/working] action create (up to date)
  * directory[create /var/opt/gitlab/gitlab-rails/tmp] action create (up to date)
  * directory[create /var/opt/gitlab/gitlab-rails/upgrade-status] action create (up to date)
  * directory[create /var/log/gitlab/gitlab-rails] action create (up to date)
  * storage_directory[/var/opt/gitlab/backups] action create
    * ruby_block[directory resource: /var/opt/gitlab/backups] action run (skipped due to not_if)
     (up to date)
  * directory[/var/opt/gitlab/gitlab-rails] action create (up to date)
  * directory[/var/opt/gitlab/gitlab-ci] action create (up to date)
  * file[/var/opt/gitlab/gitlab-rails/etc/gitlab-registry.key] action create (skipped due to only_if)
  * template[/opt/gitlab/etc/gitlab-rails/gitlab-rails-rc] action create
    - create new file /opt/gitlab/etc/gitlab-rails/gitlab-rails-rc
    - update content in file /opt/gitlab/etc/gitlab-rails/gitlab-rails-rc from none to 81d695
    --- /opt/gitlab/etc/gitlab-rails/gitlab-rails-rc	2019-01-08 20:26:23.370000000 +0000
    +++ /opt/gitlab/etc/gitlab-rails/.chef-gitlab-rails-rc20190108-30-2q4n7n	2019-01-08 20:26:23.370000000 +0000
    @@ -1 +1,3 @@
    +gitlab_user='git'
    +gitlab_group='git'
  * file[/opt/gitlab/embedded/service/gitlab-rails/.secret] action delete (up to date)
  * file[/var/opt/gitlab/gitlab-rails/etc/secret] action delete (up to date)
  * templatesymlink[Create a database.yml and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/database.yml] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/config/database.yml to /var/opt/gitlab/gitlab-rails/etc/database.yml] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/config/database.yml to /var/opt/gitlab/gitlab-rails/etc/database.yml
  
  * templatesymlink[Create a secrets.yml and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/secrets.yml] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/config/secrets.yml to /var/opt/gitlab/gitlab-rails/etc/secrets.yml] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/config/secrets.yml to /var/opt/gitlab/gitlab-rails/etc/secrets.yml
  
  * templatesymlink[Create a resque.yml and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/resque.yml] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/config/resque.yml to /var/opt/gitlab/gitlab-rails/etc/resque.yml] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/config/resque.yml to /var/opt/gitlab/gitlab-rails/etc/resque.yml
  
  * templatesymlink[Create a redis.cache.yml and create a symlink to Rails root] action create (skipped due to not_if)
  * templatesymlink[Create a redis.queues.yml and create a symlink to Rails root] action create (skipped due to not_if)
  * templatesymlink[Create a redis.shared_state.yml and create a symlink to Rails root] action create (skipped due to not_if)
  * templatesymlink[Create a smtp_settings.rb and create a symlink to Rails root] action delete
    * file[/var/opt/gitlab/gitlab-rails/etc/smtp_settings.rb] action delete (up to date)
    * link[/opt/gitlab/embedded/service/gitlab-rails/config/initializers/smtp_settings.rb] action delete (up to date)
     (up to date)
  * templatesymlink[Create a gitlab.yml and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/gitlab.yml] action create
      - update content in file /var/opt/gitlab/gitlab-rails/etc/gitlab.yml from 0607ef to 1c3614
      --- /var/opt/gitlab/gitlab-rails/etc/gitlab.yml	2019-01-08 20:24:24.056754000 +0000
      +++ /var/opt/gitlab/gitlab-rails/etc/.chef-gitlab20190108-30-129nae4.yml	2019-01-08 20:26:24.601199700 +0000
      @@ -10,7 +10,7 @@
         ## GitLab settings
         gitlab:
           ## Web server settings (note: host is the FQDN, do not include http://)
      -    host: 9187d868e738
      +    host: 67be2e823378
           port: 80
           https: false
       
      @@ -38,7 +38,7 @@
           # Uncomment and set to false if you need to disable email sending from GitLab (default: true)
           email_enabled: 
           # Email address used in the "From" field in mails sent by GitLab
      -    email_from: gitlab@9187d868e738
      +    email_from: gitlab@67be2e823378
           email_display_name: 
           email_reply_to: 
           email_subject_suffix: 
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/config/gitlab.yml to /var/opt/gitlab/gitlab-rails/etc/gitlab.yml] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/config/gitlab.yml to /var/opt/gitlab/gitlab-rails/etc/gitlab.yml
  
  * templatesymlink[Create a rack_attack.rb and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/rack_attack.rb] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/config/initializers/rack_attack.rb to /var/opt/gitlab/gitlab-rails/etc/rack_attack.rb] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/config/initializers/rack_attack.rb to /var/opt/gitlab/gitlab-rails/etc/rack_attack.rb
  
  * templatesymlink[Create a gitlab_workhorse_secret and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/gitlab_workhorse_secret] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/.gitlab_workhorse_secret to /var/opt/gitlab/gitlab-rails/etc/gitlab_workhorse_secret] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/.gitlab_workhorse_secret to /var/opt/gitlab/gitlab-rails/etc/gitlab_workhorse_secret
  
  * templatesymlink[Create a gitlab_shell_secret and create a symlink to Rails root] action create
    * template[/var/opt/gitlab/gitlab-rails/etc/gitlab_shell_secret] action create (up to date)
    * link[Link /opt/gitlab/embedded/service/gitlab-rails/.gitlab_shell_secret to /var/opt/gitlab/gitlab-rails/etc/gitlab_shell_secret] action create
      - create symlink at /opt/gitlab/embedded/service/gitlab-rails/.gitlab_shell_secret to /var/opt/gitlab/gitlab-rails/etc/gitlab_shell_secret
  
  * templatesymlink[Create a gitlab_pages_secret and create a symlink to Rails root] action create (skipped due to only_if)
  * link[/opt/gitlab/embedded/service/gitlab-rails/config/initializers/relative_url.rb] action delete (up to date)
  * file[/var/opt/gitlab/gitlab-rails/etc/relative_url.rb] action delete (up to date)
  * env_dir[/opt/gitlab/etc/gitlab-rails/env] action create
    * directory[/opt/gitlab/etc/gitlab-rails/env] action create
      - create new directory /opt/gitlab/etc/gitlab-rails/env
    * file[/opt/gitlab/etc/gitlab-rails/env/HOME] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/HOME
      - update content in file /opt/gitlab/etc/gitlab-rails/env/HOME from none to 205bb9
      --- /opt/gitlab/etc/gitlab-rails/env/HOME	2019-01-08 20:26:25.010000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-HOME20190108-30-8wplrr	2019-01-08 20:26:25.010000000 +0000
      @@ -1 +1,2 @@
      +/var/opt/gitlab
    * file[/opt/gitlab/etc/gitlab-rails/env/RAILS_ENV] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/RAILS_ENV
      - update content in file /opt/gitlab/etc/gitlab-rails/env/RAILS_ENV from none to ab8e18
      --- /opt/gitlab/etc/gitlab-rails/env/RAILS_ENV	2019-01-08 20:26:25.020000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-RAILS_ENV20190108-30-1m9v5ni	2019-01-08 20:26:25.020000000 +0000
      @@ -1 +1,2 @@
      +production
    * file[/opt/gitlab/etc/gitlab-rails/env/LD_PRELOAD] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/LD_PRELOAD
      - update content in file /opt/gitlab/etc/gitlab-rails/env/LD_PRELOAD from none to f79114
      --- /opt/gitlab/etc/gitlab-rails/env/LD_PRELOAD	2019-01-08 20:26:25.020000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-LD_PRELOAD20190108-30-1wwt5ii	2019-01-08 20:26:25.020000000 +0000
      @@ -1 +1,2 @@
      +/opt/gitlab/embedded/lib/libjemalloc.so
    * file[/opt/gitlab/etc/gitlab-rails/env/SIDEKIQ_MEMORY_KILLER_MAX_RSS] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/SIDEKIQ_MEMORY_KILLER_MAX_RSS
      - update content in file /opt/gitlab/etc/gitlab-rails/env/SIDEKIQ_MEMORY_KILLER_MAX_RSS from none to dd80d7
      --- /opt/gitlab/etc/gitlab-rails/env/SIDEKIQ_MEMORY_KILLER_MAX_RSS	2019-01-08 20:26:25.030000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-SIDEKIQ_MEMORY_KILLER_MAX_RSS20190108-30-8zgren	2019-01-08 20:26:25.030000000 +0000
      @@ -1 +1,2 @@
      +2000000
    * file[/opt/gitlab/etc/gitlab-rails/env/BUNDLE_GEMFILE] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/BUNDLE_GEMFILE
      - update content in file /opt/gitlab/etc/gitlab-rails/env/BUNDLE_GEMFILE from none to 28d586
      --- /opt/gitlab/etc/gitlab-rails/env/BUNDLE_GEMFILE	2019-01-08 20:26:25.040000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-BUNDLE_GEMFILE20190108-30-an5jd9	2019-01-08 20:26:25.040000000 +0000
      @@ -1 +1,2 @@
      +/opt/gitlab/embedded/service/gitlab-rails/Gemfile
    * file[/opt/gitlab/etc/gitlab-rails/env/PATH] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/PATH
      - update content in file /opt/gitlab/etc/gitlab-rails/env/PATH from none to d5dc07
      --- /opt/gitlab/etc/gitlab-rails/env/PATH	2019-01-08 20:26:25.050000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-PATH20190108-30-ltuifw	2019-01-08 20:26:25.050000000 +0000
      @@ -1 +1,2 @@
      +/opt/gitlab/bin:/opt/gitlab/embedded/bin:/bin:/usr/bin
    * file[/opt/gitlab/etc/gitlab-rails/env/ICU_DATA] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/ICU_DATA
      - update content in file /opt/gitlab/etc/gitlab-rails/env/ICU_DATA from none to a04260
      --- /opt/gitlab/etc/gitlab-rails/env/ICU_DATA	2019-01-08 20:26:25.060000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-ICU_DATA20190108-30-15l7b69	2019-01-08 20:26:25.060000000 +0000
      @@ -1 +1,2 @@
      +/opt/gitlab/embedded/share/icu/current
    * file[/opt/gitlab/etc/gitlab-rails/env/PYTHONPATH] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/PYTHONPATH
      - update content in file /opt/gitlab/etc/gitlab-rails/env/PYTHONPATH from none to 990cc2
      --- /opt/gitlab/etc/gitlab-rails/env/PYTHONPATH	2019-01-08 20:26:25.070000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-PYTHONPATH20190108-30-lizdi6	2019-01-08 20:26:25.070000000 +0000
      @@ -1 +1,2 @@
      +/opt/gitlab/embedded/lib/python3.4/site-packages
    * file[/opt/gitlab/etc/gitlab-rails/env/EXECJS_RUNTIME] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/EXECJS_RUNTIME
      - update content in file /opt/gitlab/etc/gitlab-rails/env/EXECJS_RUNTIME from none to 75081b
      --- /opt/gitlab/etc/gitlab-rails/env/EXECJS_RUNTIME	2019-01-08 20:26:25.080000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-EXECJS_RUNTIME20190108-30-adsyge	2019-01-08 20:26:25.080000000 +0000
      @@ -1 +1,2 @@
      +Disabled
    * file[/opt/gitlab/etc/gitlab-rails/env/TZ] action create
      - create new file /opt/gitlab/etc/gitlab-rails/env/TZ
      - update content in file /opt/gitlab/etc/gitlab-rails/env/TZ from none to 983a95
      --- /opt/gitlab/etc/gitlab-rails/env/TZ	2019-01-08 20:26:25.090000000 +0000
      +++ /opt/gitlab/etc/gitlab-rails/env/.chef-TZ20190108-30-1b7boj9	2019-01-08 20:26:25.080000000 +0000
      @@ -1 +1,2 @@
      +:/etc/localtime
  
  * link[/opt/gitlab/embedded/service/gitlab-rails/tmp] action create
    - create symlink at /opt/gitlab/embedded/service/gitlab-rails/tmp to /var/opt/gitlab/gitlab-rails/tmp
  * link[/opt/gitlab/embedded/service/gitlab-rails/public/uploads] action create
    - create symlink at /opt/gitlab/embedded/service/gitlab-rails/public/uploads to /var/opt/gitlab/gitlab-rails/uploads
  * link[/opt/gitlab/embedded/service/gitlab-rails/log] action create
    - create symlink at /opt/gitlab/embedded/service/gitlab-rails/log to /var/log/gitlab/gitlab-rails
  * link[/var/log/gitlab/gitlab-rails/sidekiq.log] action create (up to date)
  * file[/opt/gitlab/embedded/service/gitlab-rails/db/schema.rb] action create
    - change owner from 'root' to 'git'
  * remote_file[/var/opt/gitlab/gitlab-rails/VERSION] action create (up to date)
  * remote_file[/var/opt/gitlab/gitlab-rails/REVISION] action create (up to date)
  * file[/var/opt/gitlab/gitlab-rails/RUBY_VERSION] action create (up to date)
  * execute[clear the gitlab-rails cache] action nothing (skipped due to action :nothing)
  * file[/var/opt/gitlab/gitlab-rails/config.ru] action delete (up to date)
Recipe: gitlab::add_trusted_certs
  * directory[/etc/gitlab/trusted-certs] action create (up to date)
  * directory[/opt/gitlab/embedded/ssl/certs] action create (up to date)
  * file[/opt/gitlab/embedded/ssl/certs/README] action create
    - create new file /opt/gitlab/embedded/ssl/certs/README
    - update content in file /opt/gitlab/embedded/ssl/certs/README from none to 623059
    --- /opt/gitlab/embedded/ssl/certs/README	2019-01-08 20:26:25.500000000 +0000
    +++ /opt/gitlab/embedded/ssl/certs/.chef-README20190108-30-x6bg0y	2019-01-08 20:26:25.500000000 +0000
    @@ -1 +1,4 @@
    +This directory is managed by omnibus-gitlab.
    + Any file placed in this directory will be ignored
    +. Place certificates in /etc/gitlab/trusted-certs.
    - change mode from '' to '0644'
  * ruby_block[Move existing certs and link to /opt/gitlab/embedded/ssl/certs] action run

  * Moving existing certificates found in /opt/gitlab/embedded/ssl/certs

  * Symlinking existing certificates found in /etc/gitlab/trusted-certs

    - execute the ruby block Move existing certs and link to /opt/gitlab/embedded/ssl/certs
Recipe: gitlab::default
  * service[create a temporary unicorn service] action nothing (skipped due to action :nothing)
  * service[create a temporary puma service] action nothing (skipped due to action :nothing)
  * service[create a temporary sidekiq service] action nothing (skipped due to action :nothing)
  * service[create a temporary mailroom service] action nothing (skipped due to action :nothing)
Recipe: runit::sysvinit
  * execute[echo 'CS:123456:respawn:/opt/gitlab/embedded/bin/runsvdir-start' >> /etc/inittab] action run
    - execute echo 'CS:123456:respawn:/opt/gitlab/embedded/bin/runsvdir-start' >> /etc/inittab
  * execute[init q] action run
    [execute] Couldn't find an alternative telinit implementation to spawn.
    
    ================================================================================
    Error executing action `run` on resource 'execute[init q]'
    ================================================================================
    
    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of init q ----
    STDOUT: 
    STDERR: Couldn't find an alternative telinit implementation to spawn.
    ---- End output of init q ----
    Ran init q returned 1
    
    Resource Declaration:
    ---------------------
    # In /opt/gitlab/embedded/cookbooks/cache/cookbooks/runit/recipes/sysvinit.rb
    
     28: execute "init q" do
     29:   action :nothing
     30: end
    
    Compiled Resource:
    ------------------
    # Declared in /opt/gitlab/embedded/cookbooks/cache/cookbooks/runit/recipes/sysvinit.rb:28:in `from_file'
    
    execute("init q") do
      action [:nothing]
      default_guard_interpreter :execute
      command "init q"
      backup 5
      returns 0
      user nil
      declared_type :execute
      cookbook_name "runit"
      recipe_name "sysvinit"
      domain nil
    end
    
    System Info:
    ------------
    chef_version=13.6.4
    platform=ubuntu
    platform_version=16.04
    ruby=ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]
    program_name=/opt/gitlab/embedded/bin/chef-client
    executable=/opt/gitlab/embedded/bin/chef-client
    
Recipe: gitlab::gitlab-rails
  * execute[clear the gitlab-rails cache] action run (skipped due to not_if)

Running handlers:
There was an error running gitlab-ctl reconfigure:

execute[init q] (runit::sysvinit line 28) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of init q ----
STDOUT: 
STDERR: Couldn't find an alternative telinit implementation to spawn.
---- End output of init q ----
Ran init q returned 1

Running handlers complete
Chef Client failed. 43 resources updated in 17 seconds
```

**Expected behavior**
It should work like VM version.

`docker --version`
Docker version 18.09.0, build 4d60db4

`systeminfo | findstr /C:"OS" | clip`
``` log
OS Name:                   Microsoft Windows 10 Pro Insider Preview
OS Version:                10.0.18309 N/A Build 18309
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
BIOS Version:              LENOVO 39CN16WW       , 7/29/2010
```

**Comment by n3rd4i** [comment]
strace of the issue [here](https://gist.github.com/n3rd4i/83adc29ed88f4824cb5d6b8afec0dfda)

**Comment by n3rd4i** [comment]
After doing some googling I found an workaround fix [here](https://gitlab.com/gitlab-org/omnibus-gitlab/issues/3995#note_126635237), for the previous error...
* `[...] Have to touch /.dockerenv before starting [...]`

I used `docker-compose up` instead of `docker run` commands:
* link to the [docker-compose.yaml](https://gist.github.com/n3rd4i/7e2299378c91091216a218e167b4ce91)

This creates a new result:
``` log
web_1  | Multiple failures occurred:
web_1  | * Errno::ENOMEM occurred in delayed notification: execute[clear the gitlab-rails cache] (gitlab::gitlab-rails line 398) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart redis svlogd configuration] (redis::enable line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart postgresql svlogd configuration] (postgresql::enable line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart unicorn svlogd configuration] (gitlab::unicorn line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart sidekiq svlogd configuration] (gitlab::sidekiq line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[gitlab-workhorse] (gitlab::gitlab-workhorse line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart gitlab-workhorse svlogd configuration] (gitlab::gitlab-workhorse line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart nginx svlogd configuration] (nginx::enable line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart logrotate svlogd configuration] (gitlab::logrotate line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[gitaly] (gitaly::enable line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart gitaly svlogd configuration] (gitaly::enable line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[node-exporter] (gitlab::node-exporter line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart node-exporter svlogd configuration] (gitlab::node-exporter line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart gitlab-monitor svlogd configuration] (gitlab::gitlab-monitor line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[redis-exporter] (gitlab::redis-exporter line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart redis-exporter svlogd configuration] (gitlab::redis-exporter line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[prometheus] (gitlab::prometheus line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart prometheus svlogd configuration] (gitlab::prometheus line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[alertmanager] (gitlab::alertmanager line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart alertmanager svlogd configuration] (gitlab::alertmanager line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: service[postgres-exporter] (gitlab::postgres-exporter line 227) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)
web_1  | * Errno::ENOMEM occurred in delayed notification: ruby_block[restart postgres-exporter svlogd configuration] (gitlab::postgres-exporter line 99) had an error: Errno::ENOMEM: Cannot allocate memory - fork(2)

```
* [full log](https://gist.github.com/n3rd4i/0c64260f7d5dcecb96464b7c07bb803d) of the new issue

All in all, WSL-based installation creates a different behavior than Real Linux.

**Comment by n3rd4i** [comment]
probably this works in WSL2, closing

**Comment by fjakop** [comment]
This is not working in WSL 2 on Win 10 Enterprise backed by Ubuntu 18.04.

**Comment by zckman** [comment]
> `STDERR: chmod: changing permissions of '/var/opt/gitlab/.ssh': Operation not permitted`

Same issue for me on Windows 10 with WSL2 and these distros:

- Alpine using Docker installed via package manager
- Fedora using WSL integrated Docker

**Comment by Biswa96** [comment]
@zckman Can you provide steps/commands to reproduce the issue?

**Comment by zckman** [comment]
> @zckman Can you provide steps/commands to reproduce the issue?

I use [yuk7/AlpineWSL](https://github.com/yuk7/AlpineWSL). Docker was installed with apk.

```
λ wsl --version
WSL-Version: 1.0.3.0
Kernelversion: 5.15.79.1
WSLg-Version: 1.0.47
MSRDC-Version: 1.2.3575
Direct3D-Version: 1.606.4
DXCore-Version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows Version: 10.0.19044.2364

λ wsl -l -v
  NAME                   STATE           VERSION
* Alpine                 Running         2
```

This is my `docker-compose.yml`:

```yaml
version: '3'
services:
  gitlab:
    image: gitlab/gitlab-ce:15.6.0-ce.0
    environment:
      GITLAB_OMNIBUS_CONFIG: |
        external_url 'http://localhost:52080'
    volumes:
      - ./data/gitlab/config:/etc/gitlab:Z
      - ./data/gitlab/logs:/var/log/gitlab:Z
      - ./data/gitlab/data:/var/opt/gitlab:Z
    ports:
      - 52080:80
      - 52443:443
      - 52022:22
```

When I enter the WSL shell and run `docker-compose up`, I get this and the container exits:
`STDERR: chmod: changing permissions of '/var/opt/gitlab/.ssh': Operation not permitted`

**Comment by Biswa96** [comment]
I get a different error message. Am I doing anything wrong here?

```
$ docker-compose up
Creating test_gitlab_1 ... error
ERROR: for test_gitlab_1  Cannot create container for service gitlab: invalid port specification: "3121342"
ERROR: for gitlab  Cannot create container for service gitlab: invalid port specification: "3121342"
ERROR: Encountered errors while bringing up the project.
```

**Comment by Biswa96** [comment]
I found that the ports need to be in double quote in that docker-compose.yml file. After doing that, things are working fine. There are some logs but no error so far.

**Comment by zckman** [comment]
I guess you're about the quotes. But the container exits before that.

Here are the logs:
[docker-compose-up.out.txt](https://github.com/microsoft/WSL/files/10298433/docker-compose-up.out.txt)
[cinc-stacktrace.out.txt](https://github.com/microsoft/WSL/files/10298434/cinc-stacktrace.out.txt)

**Comment by Biswa96** [comment]
I am not sure why it causes error in your setup. But that docker image is working fine in my system. I have not seen any error in the docker output. Though the memory usage just ramping up more than 4 GB. Running GitLab instance in WSL2 may not be a proper solution.

**Comment by Biswa96** [comment]
BTW, I do not use Docker Desktop for Windows, only dockerd and docker command in Linux distribution. Just preference :)

**Comment by yirami** [comment]
Same error occurred when I use `-v` to bind a windows  path (/mnd/d/balabala) for `/var/opt/gitlab`. After google, I found this [coment](https://stackoverflow.com/a/66450698/19527989). Then I try to use command `docker volume create xxx` to create a volume and bind it to `/var/opt/gitlab`. The error disappeared.

PS:
+ Windows 10.0.19044.2251
+ WSL2 with Debian
+ Docker for windows 4.14.1 (91661)
